README.md
![Meinter Logo](vignettes/figs/meinter.png)

MeinteR (MEthylation INTERpretation) is an R package that identifies critical differentially methylated sites, based on the following hypothesis: Critical methylation-mediated changes are more likely in genomic regions enriched in cis-acting regulatory elements than in genomic “deserts”. MeinteR calculates the abundance of co-localized elements, such as transcription factor binding sites, tentative splice sites, and other DNA features, such as G-quadruplexes and palindromes that potentially lead to distinct DNA conformational features and rank them with respect to their putative methylation impact.

---

## Package files

[Binary Package](dist/MeinteR_0.99.0.tgz)

[Source Package](dist/MeinteR_0.99.0.tar.gz)

[Package Vignette (PDF)](vignettes/Meinter_vignette.pdf)

[Package Manual (PDF)](vignettes/MeinteR.pdf) 



## Installation

To install MeinteR and all its dependencies install and load devtools. 

**Local installation ** 
First, download the binary distribution of the package, unzip it to a local folder and run the following commands:
```
library(devtools)
#Set folder of the package
package.folder <- "~/MeinteR"
install_local(package.folder)
```
**Install from GitHub**

MeinteR and its dependencies can be directly installed from GitHub using the following commands:
```
library(devtools)
devtools::install_gihub("andigoni/meinter")
```

**Source code for use cases**

[Use Case 1: Genome-wide association of G-quadruplexes with DNA methylation using TCGA/GEO breast cancer datasets](vignettes/UseCase1.Rmd)

[Use Case 2: Evaluation of genomic signatures on cancer methylation profiles](vignettes/UseCase2.Rmd)

[Use Case 3: Associating genomic DMS signatures with gene expression](vignettes/UseCase3.Rmd)



**Demo run**

```
library(MeinteR)
rm(list = ls())
#Reorder columns of sample dataset
re.sample <- reorderBed(sample, 1, 2, 3, 5, 4)
#Select sites with delta-beta > 0.50
bed.data <- re.sample[re.sample$score >= 0.50,] 
#Find overlaps with alternative splicing events
altSS <- findAltSplicing(bed.data)
#Find overlaps with splice sites
ss <- findSpliceSites(bed.data, persim = 0.8, offset = 10)
#Find palindromic sequences
pals <- findPals(bed.data)
#Find G-quadruplexes
quads <- findQuads(bed.data, offset = 50)
#Find overlaps with "MA0107.1", "MA0098", "MA115.1", "MA0131.2 transcription factors
tfbs <-
  findTFBS(
    bed.data,
    target = "all",
    tf.ID = c("MA0107.1", "MA0098", "MA115.1", "MA0131.2")
  )
#Find overlaps with conserved transcription factors. ~71MB gzipped file 
#will be downloaded if local path is not set.
ctfbs <- findConservedTFBS(bed.data)
#Find overlaps with DNA shape features
shapes <- findShapes(bed.data)
#Feature weights
weights = list()
weights[["spls"]] = 1
weights[["ctfbs"]] = 1
weights[["tfbs"]] = 1
weights[["pals"]] = 1
weights[["quads"]] = 1
weights[["shapes"]] = 1
#Mapping MeinteR arguments with function outputs
funList = list()
funList[["spls"]] = ss
funList[["altss"]] = altSS
funList[["tfbs"]] = tfbs
funList[["ctfbs"]] = ctfbs
funList[["pals"]] = pals
funList[["quads"]] = quads
funList[["shapes"]] = shapes
#Calculate genomic index
index <- meinter(re.sample, funList, weights)

```

**Citation**

A. Malousi et al. "MeinteR: A computational method to prioritize aberrant DNA methylation events using local genomic substrate" (submitted)







inst/NEWS
MeinteR is an R package that can be used to prioritize aberrant DNA methylation events that are identified in microarray and next-generation sequencing platforms.

tests/testthat.R
library(testthat)
library(MeinteR)

test_check("MeinteR")


tests/testthat/test_packages.R
rm(list=ls())
packages <- c("ggplot2", "parallel", "plyr","stats4","reshape2", "pqsfinder","BSgenome.Hsapiens.UCSC.hg19", "GenomicRanges", 
              "BSgenome","tidyverse","Biostrings","XVector","GenomeInfoDb","S4Vectors","stats4", "stats","IRanges","BiocGenerics",
              "parallel","base","rtracklayer","reshape2","TFBSTools","DNAshapeR","JASPAR2018" )
new.packages <- packages[!(packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

bioconductor.packages <- c("BSgenome.Hsapiens.UCSC.hg19","GenomicRanges","rtracklayer", "BiocGenerics","S4Vectors", "IRanges","GenomeInfoDb","BSgenome","Biostrings","XVector")
bioc.packages <- bioconductor.packages[!(bioconductor.packages  %in% installed.packages()[,"Package"])]
if (length(bioc.packages)) {
  source("http://bioconductor.org/biocLite.R")
  biocLite(bioc.packages)
}

R/shapes.R
#######################################################################
## Package    : MeinteR
## File       : shapes.R
## Functions  : findShapes (exported)
##
## Updated    : 30-04-2019
##
## Title      : Function for the detection of DNA shape features
#######################################################################


#' Find putative conformational DNA changes
#'
#' Predicts conformational changes of DNA shapes, such as
#' minor groove width (MGW), Roll, propeller twist (ProT) and helix twist (HeIT)
#' in the unmethylated and methylated context using methyl-DNAshape algorithm.
#'
#' @param bed.data A data frame containing input bed-formatted data
#' @param offset Number of nucleotides expanded in each direction (default:50, max:200)
#' @return 1/ p-value of the MGW in the unmethylated/methylated CpG context for each sequence
#' @return 2/ p-value of the HeIT in the unmethylated/methylated CpG context for each sequence
#' @return 3/ p-value of the ProT in the unmethylated/methylated CpG context for each sequence
#' @return 4/ p-value of the Roll in the unmethylated/methylated CpG context for each sequence
#' @export

findShapes <- function(bed.data, offset = 50) {
  if (offset > 200 | offset < 0) {
    stop('Valid offset: [1..200]')
  }
  if (sum(is.na(bed.data)) > 0) {
    stop("Input dataset contains ", sum(is.na(bed.data)), " missing values.")
  }
  bed.data$strand[is.na(bed.data$strand)] <- "+"
  if (mean(bed.data$end - bed.data$start) <= 1) {
    adjust.C <- 0
  } else {
    adjust.C <- 1
  }
  if (mean(bed.data$end - bed.data$start) > 3) {
    exit("The width of the methylation bed data should be less than 4nt.")
  }
  bed.data$chr <-
    droplevels(bed.data$chr) # Drop levels not observed in the dataset
  gr <-
    GRanges(
      seqnames = bed.data$chr,
      ranges = IRanges(
        start = bed.data$start - offset + adjust.C,
        width = 2 * offset + 1
      )
    )
  
  fn_methy <-
    system.file("extdata", "MethylSample.fa", package = "DNAshapeR")
  suppressMessages(getFasta(
    gr,
    Hsapiens,
    width = 2 * offset + 1,
    filename = fn_methy
  ))
  fap = fn_methy
  m = vector()
  meth.pos = offset + 1
  for (i in seq_len(length(gr))) {
    x = paste0(">seq", i)
    m = rbind(m, x)
    m = rbind(m, meth.pos)
  }
  fn_methy_pos <-
    system.file("extdata", "MethylSamplePos.fa", package = "DNAshapeR")
  
  write.table(
    m,
    file = fn_methy_pos,
    col.names = FALSE,
    row.names = FALSE,
    quote = FALSE
  )
  pred <- getShape(fn_methy)
  pred_methy <-
    getShape(fn_methy,
             methylate = TRUE,
             methylatedPosFile = fn_methy_pos)
  #Statistical evaluation
  wt.mgw <- wt.HelT <- wt.ProT <- wt.Roll <- c()
  for (i in 1:nrow(bed.data)) {
    wt <-
      wilcox.test(x = pred$MGW[i, ],
                  y = pred_methy$MGW[i, ],
                  paired = TRUE)
    wt.mgw <- c(wt.mgw, wt$p.value)
    wt <-
      wilcox.test(x = pred$HelT[i, ],
                  y = pred_methy$HelT[i, ],
                  paired = TRUE)
    wt.HelT <- c(wt.HelT, wt$p.value)
    wt <-
      wilcox.test(x = pred$ProT[i, ],
                  y = pred_methy$ProT[i, ],
                  paired = TRUE)
    wt.ProT <- c(wt.ProT, wt$p.value)
    wt <-
      wilcox.test(x = pred$Roll[i, ],
                  y = pred_methy$Roll[i, ],
                  paired = TRUE)
    wt.Roll <- c(wt.Roll, wt$p.value)
  }
  bed.coor <-
    data.frame(paste(bed.data$chr, bed.data$start, bed.data$end, sep = "_"))
  p.MGW <- cbind(bed.coor, wt.mgw)
  colnames(p.MGW) <- c("seqname", "p.MGW")
  p.HelT <- cbind(bed.coor, wt.HelT)
  colnames(p.HelT) <- c("seqname", "p.HelT")
  p.ProT <- cbind(bed.coor, wt.ProT)
  colnames(p.ProT) <- c("seqname", "p.ProT")
  p.Roll <- cbind(bed.coor, wt.Roll)
  colnames(p.Roll) <- c("seqname", "p.Roll")
  
  par(mfrow = c(4, 2))
  plotShape(pred$MGW, main = "Unethylated Minor Groove Width")
  panel.first = grid(equilogs = FALSE)
  plotShape(pred_methy$MGW, main = "Methylated Minor Groove Width", colLine = 'red')
  panel.first = grid(equilogs = FALSE)
  plotShape(pred$HelT, main = "Unethylated Helix Twist")
  panel.first = grid(equilogs = FALSE)
  plotShape(pred_methy$HelT, main = "Methylated Helix Twist",  colLine = 'red')
  panel.first = grid(equilogs = FALSE)
  plotShape(pred$ProT, main = "Unethylated Propeller Twist")
  panel.first = grid(equilogs = FALSE)
  plotShape(
    pred_methy$ProT,
    main = "Methylated Propeller Twist",
    ylim = c(-7, -5.5),
    colLine = 'red'
  )
  panel.first = grid(equilogs = FALSE)
  plotShape(pred$Roll, main = "Unethylated Roll")
  panel.first = grid(equilogs = FALSE)
  plotShape(
    pred_methy$Roll,
    main = "Methylated Roll",
    ylim = c(-2, 4),
    colLine = 'red'
  )
  panel.first = grid(equilogs = FALSE)
  results = list()
  results[[1]] <- p.MGW
  results[[2]] <- p.HelT
  results[[3]] <- p.ProT
  results[[4]] <- p.Roll
  return(results)
}


vignettes/demoRun.R

#Demo run
# Input: sample dataset
library(MeinteR)
rm(list = ls())
#Reorder columns
re.sample <- reorderBed(sample, 1, 2, 3, 5, 4)
#Select site with |delta-beta|>0.5
bed.data <- re.sample[re.sample$score >= 0.50,] 
#Core function calling
altSS <- findAltSplicing(bed.data)
ss <- findSpliceSites(bed.data, persim = 0.8, offset = 10)
pals <- findPals(bed.data)
quads <- findQuads(bed.data, offset = 50)
tfbs <-
  findTFBS(
    bed.data,
    target = "all",
    tf.ID = c("MA0107.1", "MA0098", "MA115.1", "MA0131.2")
  )
ctfbs <-
  findConservedTFBS(bed.data, known.conserved.tfbs.file = "~/Downloads/tfbsConsSites.txt.gz")
shapes <- findShapes(bed.data)
#Weight assignment
weights = list()
weights[["spls"]] = 1
weights[["ctfbs"]] = 1
weights[["tfbs"]] = 1
weights[["pals"]] = 1
weights[["quads"]] = 1
weights[["shapes"]] = 1
#Mappings to function lists
funList = list()
funList[["spls"]] = ss
funList[["altss"]] = altSS
funList[["tfbs"]] = tfbs
funList[["ctfbs"]] = ctfbs
funList[["pals"]] = pals
funList[["quads"]] = quads
funList[["shapes"]] = shapes
#Calculate genomic index
index <- meinter(re.sample, funList, weights)


R/misc.R
#############################################################################
## Package    : MeinteR
## File       : misc.R
## Functions  : loadFile (exported)
##            : bed2Seq (exported)
##            : nameStudy (exported)
##            : isEmptyDF
##            : reorderBed (exported)
##            : filterByProm
##            : filterByCGI
##            : is.defined
##            : validateBed
##            : plotCpG (exported)
##            : plotBeta (exported)
##            : importGEO  (exported)
##            : loadAnnotationGEO
##            : validateGEO
##            : groupGEO
##            : probesChr
##            : loadSeqGEO (exported)
##            : mval2beta
## Updated    : 30-04-2019
##
## Title      : Miscellaneous functions
#############################################################################


rm(list = ls())
#Package local environmental variables
pkg.env <- new.env()
local(NAME <- "study", envir = pkg.env)

#Install package dependencies
packages <-
  c(
    "ggplot2",
    "plyr",
    "stats4",
    "reshape2",
    "tidyverse",
    "XVector",
    "GenomeInfoDb",
    "S4Vectors",
    "stats",
    "IRanges",
    "BiocGenerics",
    "parallel",
    "base",
    "reshape2",
    "grDevices",
    "graphics",
    "utils"
  )
new.packages <-
  packages[!(packages %in% installed.packages()[, "Package"])]
if (length(new.packages))
  install.packages(new.packages)

bioconductor.packages <-
  c(
    "TFBSTools",
    "TxDb.Hsapiens.UCSC.hg19.knownGene",
    "Biostrings",
    "BSgenome",
    "pqsfinder",
    "Biobase",
    "DNAshapeR",
    "JASPAR2018",
    "BSgenome.Hsapiens.UCSC.hg19",
    "GenomicRanges",
    "rtracklayer",
    "BiocGenerics",
    "S4Vectors",
    "JASPAR2018",
    "IRanges",
    "pqsfinder",
    "GenomeInfoDb",
    "BSgenome",
    "GenomicFeatures",
    "XVector",
    "GEOquery",
    "FDb.InfiniumMethylation.hg19"
  )
bioc.packages <-
  bioconductor.packages[!(bioconductor.packages  %in% installed.packages()[, "Package"])]
if (length(bioc.packages)) {
  source("http://bioconductor.org/biocLite.R")
  biocLite(bioc.packages)
}



#'Load input data
#'
#' Loads tabular files containing methylation data. The function checks the delimiter and validates the order
#' of the columns (chr, start,end,score,strand).
#'
#' @param   FH   Full path of the tabular methylation data
#' @return  df   A data frame with the tabular methylation data
#' @export

loadFile <- function(FH) {
  title.line <- readLines(FH, n = 1)
  if (grepl(";", title.line)) {
    bed.data <- read.csv(file = FH,
                         header = TRUE,
                         sep = ";")
  }
  if (grepl(",", title.line)) {
    bed.data <- read.csv(file = FH,
                         header = TRUE,
                         sep = ",")
  }
  if (grepl(",\t", title.line)) {
    bed.data <- read.csv(file = FH,
                         header = TRUE,
                         sep = "\t")
  }
  message("Valid column names and ordering: chr,start,end,score,strand")
  message("Use reorderbed function to change the column names and order.")
  return(bed.data)
}

#'Fetch sequences from bed-formatted data frames
#' @param bedline  Valid bed-formatted data frame
#' @param offset Number of nucleotides expanded in each direction ([1,1000])
#' @return A DNAStringset containing the sequences in hg19 genome assembly
#' @export
#'
bed2Seq <- function(bedline, offset) {
  #Depends on Biostrings, BSgenome.Hsapiens.UCSC.hg19
  if (!(dplyr::between(offset, 0, 1000)))
    stop("Offset should be between 1 and 1000.")
  if (sum(is.na(bedline)) > 0) {
    stop("Input dataset contains ", sum(is.na(bedline)), "missing values.")
  }
  strands = c("+", "-")
  if (!exists("bedline$strand")) {
    bedline$strand = "+"
  }
  seq <- getSeq(
    Hsapiens,
    bedline$chr ,
    start = bedline$start - offset,
    end = bedline$end + offset,
    strand = bedline$strand
  )
  seq@ranges@NAMES = paste0(bedline$chr,
                            sep = "_",
                            bedline$start,
                            sep = "_",
                            bedline$end)
  return(seq)
}


#'Set a study name
#'
#'Sets a name to the analysis that appears in the exported plots.
#'
#' @param study.name  A string corresponding to the name of the study
#' @return The name of the study
#' @export
nameStudy <- local(function (study.name) {
  NAME <<- study.name
  return(NAME)
}, envir = pkg.env)


#'Check if data frame is empty
#'
#'Checks if a data frame has no values
#'
#' @param df  The input data frame
#' @return TRUE/FALSE (TRUE is the data frame is empty)
#' @export
isEmptyDF <- function(df) {
  flag <- FALSE
  if (nrow(df) == 0) {
    flag <- TRUE
  }
  return(flag)
}


#' Reorder tabular methylation data to bed format
#'
#' Reorders tabular methylation data to bed-formatted files.
#' Compatible inputs are .txt, .csv data and other textual formats that
#' contain the following mandatory columns: chr, start, end and score.
#'
#'
#' @param input.data A data frame containing input bed-formatted data
#' @param chr.col Column number containing the chromosome name
#' @param start.col Column number containing the chromosome's start position
#' @param end.col Column number containing the chromosome's end position
#' @param score.col Column number containing the methylation score values either beta or delta-beta
#' @param strand.col Column number containing the strand in the use data file ('+' strand is assumed if strand column is missing)
#' @return A valid bed-formatted file (input of the `MeinteR::find*` functions)
#' @export
reorderBed <-
  function(input.data,
           chr.col,
           start.col,
           end.col,
           score.col,
           strand.col = NULL) {
    if (!is.defined(strand.col)) {
      input.data$strand = "+"
      bed <-
        cbind(input.data[, c(chr.col, start.col, end.col, score.col)], input.data$strand)
    } else {
      bed <-
        cbind(input.data[, c(chr.col, start.col, end.col, score.col, strand.col)])
    }
    colnames(bed) = c("chr", "start", "end", "score", "strand")
    bed$start <- as.numeric(as.character(bed$start))
    bed$end <- as.numeric(as.character(bed$end))
    bed$score <- as.numeric(as.character(bed$score))
    return (bed)
  }

#' Filter by promoters
#'
#' Selects genomic coordinates included in promoters based on the UCSC hg19 gene coordinates.
#'
#' @param input.data  A data frame containing input data in bed format
#' @param up.tss  Number of nucleotides upstream transcription start site
#' @param down.tss  Number of nucleotides downstream transcription start site
#' @return A data frame with the CpG sites located in promoter regions

filterByProm <- function(input.data, up.tss, down.tss) {
  if (sum(is.na(input.data)) > 0) {
    stop("Input dataset contains ", sum(is.na(input.data)), " missing values.")
  }
  txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
  genes <- transcriptsBy(txdb, "gene")
  proms <-
    promoters(genes, upstream = up.tss, downstream = down.tss) #IRanges object
  input.GRanges <-
    makeGRangesFromDataFrame(input.data) #GRanges object
  x <-
    as.data.frame(subsetByOverlaps(input.GRanges, proms, type = "within"))
  res <-
    merge(
      input.data,
      x,
      by.x = c("chr", "start", "end", "strand"),
      by.y = c("seqnames", "start", "end", "strand")
    )
  res <- reorderBed(res, 1, 2, 3, 5, 4)
  return(res)
}

#' Filter by CpG islands
#'
#' Selects genomic coordinates included in CpG islands, using the cpgIslands dataset.
#'
#' @param input.data  A data frame containing input data in bed format
#' @return A data frame with the CpG sites located in CpG islands


filterByCGI <- function(input.data) {
  if (sum(is.na(input.data)) > 0) {
    stop("Input dataset contains ", sum(is.na(input.data)), " missing values.")
  }
  CGI <- makeGRangesFromDataFrame(cpgIslands)
  input.GRanges <-
    makeGRangesFromDataFrame(input.data) #GRanges object
  x <-
    as.data.frame(subsetByOverlaps(input.GRanges, CGI, type = "within"))
  res <-
    merge(
      input.data,
      x,
      by.x = c("chr", "start", "end", "strand"),
      by.y = c("seqnames", "start", "end", "strand")
    )
  res <- reorderBed(res, 1, 2, 3, 5, 4)
  return(res)
}


is.defined <- function(x)
  ! is.null(x)


#'Validate format of the input bed data
#'
#'Validates input methylation data. Checks the presence of the chr, start, end, score columns.
#'If column `strand` is not set then `+` strand is assumed.
#'Cleans rows with empty cells and sets numeric format to the start, end and score columns.
#'
#' @param bed.data  A data frame containing input bed-formatted data
#' @param omit.na Omit rows with empty cells (default:TRUE)
#' @return A well-formatted data frame
#' @export
#'
validateBed <- function(bed.data, omit.na = TRUE) {
  if (isEmptyDF(bed.data)) {
    stop("Empty data frame")
  }
  valid.cols <- c("chr", "start", "end", "score", "strand")
  if (!all(colnames(bed.data) %in% (valid.cols))) {
    stop("Invalid column names. Use reorderBed to match bed data with the column names.")
  }
  strands = c("+", "-")
  if (is.null(bed.data$strand)) {
    bed.data$strand <- "+"
  }
  if (isFALSE(bed.data$strand %in% (strands))) {
    bed.data$strand <- "+"
  }
  if (omit.na) {
    bed.data <- na.omit(bed.data)
  }
  if (unique(tolower(substring(bed.data$chr, 1, 3)) != "chr")) {
    bed.data$chr <- paste0("chr", bed.data$chr)
  }
  bed.data$start <- as.numeric(bed.data$start)
  bed.data$end <- as.numeric(bed.data$end)
  bed.data$score <- as.numeric(bed.data$score)
  return(bed.data)
}



#'Plot G+C-content and observed/expected ratio.
#
#'
#' Generates density plots of the G+C content and
#' observed/expected CpG ratio for the input dataset and the human genome CpG islands
#'
#' @param bed.data  A data frame containing input bed-formatted data
#' @param offset  Number of nucleotides expanded in each direction (default:200, min:20 max:1000)
#' @return 1/ A data frame containing the G+C content (percentage of island that is C or G)
#' and ratio of observed (CpG number) to expected(Number of C* Number of G/sequence length)
#' @return 2/ Density plot of the G+C-content
#' @return 3/ Density plot of the observed/expected ratio
#' @export

plotCpG <- function(bed.data, offset = 200) {
  if (!(dplyr::between(offset, 20, 1000)))
    stop("Check offset value (valid range [20,1000]).")
  while (!is.null(dev.list()))
    dev.off()
  seqs = bed2Seq(bed.data, offset)
  res = alphabetFrequency(seqs)
  L = rowSums(res)
  cprop = res[, "C"]
  gprop = res[, "G"]
  GC = (cprop + gprop) / L
  denom = cprop * gprop
  nom = vcountPattern("CG", seqs) * L
  Observed = nom / denom
  median(Observed)
  cpgIslands$obsExp <- as.numeric(as.character(cpgIslands$obsExp))
  cpgIslands$perGc <- as.numeric(as.character(cpgIslands$perGc))
  d.gc.ref <-
    density(cpgIslands$perGc / 100, na.rm = TRUE, kernel = "gaussian")
  d.gc <- density(GC, na.rm = TRUE, kernel = "gaussian")
  y.max <- max(max(d.gc$y), max(d.gc.ref$y))
  plot(
    d.gc.ref,
    xlab = "G+C-content",
    main = "Density of the G+C-content",
    xlim = c(0, 1),
    ylim = c(0, y.max),
    sub = paste0("OFFSET:", offset, "nt, N:", nrow(bed.data))
  )
  grid (NULL, NULL, lty = 6, col = "lightgrey")
  polygon(d.gc.ref, col = 'gold', border = "black")
  polygon(d.gc, col = 'lightskyblue4')
  abline(
    v = 0.5,
    b = 0,
    col = "darkgrey",
    lwd = 3,
    lty = 2
  )
  legend(
    "topright",
    legend = c("CpG islands (hg19)", local(NAME, envir = pkg.env)),
    fill = c('gold', 'lightskyblue4')
  )
  d.obs.ref <-
    density(cpgIslands$obsExp, na.rm = TRUE, kernel = "gaussian")
  d.obs <- density(Observed, na.rm = TRUE, kernel = "gaussian")
  y.max <- max(max(d.obs$y), max(d.obs.ref$y))
  plot(
    d.obs.ref,
    xlab = "Observed/Expected CpG ratio",
    main = "Density of the observed/expected CpG ratios",
    xlim = c(0, 2),
    ylim = c(0, y.max),
    sub = paste0("OFFSET: ", offset, "nt, N: ", nrow(bed.data))
  )
  grid (NULL, NULL, lty = 6, col = "lightgrey")
  polygon(d.obs.ref, col = 'gold', border = "black")
  polygon(d.obs, col = 'lightskyblue4')
  abline(
    v = 0.6,
    b = 0,
    col = "darkgrey",
    lwd = 3,
    lty = 2
  )
  legend(
    "topright",
    legend = c("CpG islands (hg19)", local(NAME, envir = pkg.env)),
    fill = c('gold', 'lightskyblue4')
  )
  res <- data.frame(GC, Observed)
  return(res)
}


#'Plot scores of the input data
#
#' Generates a density plot of the score values listed in the input dataset.
#'
#' @param bed.data A data frame containing input bed-formatted data
#' @export
plotBeta <- function(bed.data) {
  if ("score" %in% colnames(bed.data)) {
    if (!is.numeric(bed.data$score)) {
      stop("Column score must be numeric.")
    }
    if (sum(is.na(bed.data)) > 0) {
      stop("Input dataset contains ", sum(is.na(bed.data)), " missing values.")
    }
    dens <- density(bed.data$score, na.rm = TRUE, kernel = "gaussian")
    plot(dens,
         xlab = "Methylation level",
         main = "Distribution of the methylation values")
    grid (NULL, NULL, lty = 6, col = "lightgrey")
    polygon(dens, col = 'lightskyblue4')
    legend("topright",
           legend =  local(NAME, envir = pkg.env),
           fill = 'lightskyblue4')
  } else {
    stop("Score column does not exist in the input data.")
  }
}



#' Import GEO data series in the workspace
#'
#' Imports GEO data series. The function fetches data matrices corresponding to a pre-defined GSE identifier
#' and builds valid, bed-formatted dataset with delta-beta values between two sample groups, described
#' in a user-defined annotation file.
#'
#' @param gse.acc   A string corresponding to the accession number of the GEO data series
#' @param annotation.file A string corresponding to the full local path to the annotation files containing
#' sample grouping information
#' @return 1/ A bed-formatted data frame with the chromosomal coordinates and of each methylation probe and
#' the corresponding delta-beta values between the two groups
#' @return 2/ Beta values of each sample listed in the annotation file
#' @return 3/ Annotation data frame
#' @return 4/ Mean beta values of group 1
#' @return 5/ Mean beta values of group 2
#' @name importGEO
#' @export
importGEO <- function(gse.acc, annotation.file) {
  if (!is.defined(gse.acc)) {
    stop("FAIL: Data series accession number is missing.")
  }
  if (!is.defined(annotation.file)) {
    stop("Sample annotation is missing.")
  }
  if (grepl("^GSE\\d\\d+$", gse.acc)) {
    message("PASS: Valid accession number for the data series.")
  } else
  {
    stop("Invalid accession number. Currently, GSE accession numbers are supported.")
  }
  anno.df <- loadAnnotationGEO(annotation.file)
  gse <- getGEO(gse.acc, GSEMatrix = TRUE, getGPL = TRUE)
  if (length(gse) > 1) {
    stop(
      "MeinteR currently supports single-matrix data series. Data series ",
      gse.acc,
      " has ",
      length(gse),
      "."
    )
  }
  if (any(is.na(gse[[1]]@featureData@data$ID))) {
    stop("Missing probe identifiers. Retry to download data series.")
  }
  valid <- validateGEO(gse)
  if (valid) {
    "Data series looks good."
  } else
  {
    stop("Failed to import data series in the workspace.")
  }
  if (!all(anno.df$sample %in% gse[[1]]$geo_accession)) {
    stop(
      "Inconsistency between the accession numbers of the annotation file and the GEO data series.
      Check correspondence of the GSM IDs and non-visible characters in the annotation file."
    )
  }
  message("Data series: ", length(gse[[1]]$geo_accession), " samples.")
  message("Annotation file: ", nrow(anno.df), " samples.")
  all.betas <- as.data.frame(exprs(gse[[1]]))
  betas <- all.betas[, anno.df[, 1]]
  if (min(betas, na.rm = TRUE) < 0 |
      max(betas, na.rm = TRUE) > 1) {
    message("M-values are detected and will be transformed to beta values.")
    betas <- mval2beta(betas)
  }
  grp <- groupGEO(anno.df, betas)
  probe.d.beta <- grp[[1]]
  res <- probesChr(gse, probe.d.beta)
  result = list()
  result[[1]] <- res
  result[[2]] <- betas
  result[[3]] <- anno.df
  result[[4]] <- grp[[2]]
  result[[5]] <- grp[[3]]
  return(result)
  }

#Internal function for loading annotation files
loadAnnotationGEO <- function(file.path) {
  if (!file.exists(file.path)) {
    stop(paste0("File ", file.path, " not found. Check full path."))
  }
  valid.cols <- c("SAMPLE", "STATUS")
  anno <-
    read.csv(file = file.path,
             header = TRUE,
             stringsAsFactors = FALSE)
  #Remove leading and/or trailing whitespace from character strings.
  anno <-
    data.frame(lapply(anno, trimws), stringsAsFactors = FALSE)
  if (all(toupper(colnames(anno)) %in% valid.cols)) {
    message("PASS: Columns SAMPLE and STATUS are present.")
  } else {
    stop("Column names <samples> and <status> are not detected in the first line.")
  }
  if (length(colnames(anno)) != 2) {
    stop("Annotation file columns: sample identifier (GSM accession number), sample status")
  }
  if (length(unique(anno$status)) != 2) {
    stop(
      "Other than two conditions are detected. Check non-visible characters in the annotation file."
    )
  }
  return(anno)
}


validateGEO <- function(gse.list) {
  valid = TRUE
  #Validate species
  SPECIES = "9606"
  if (table(gse.list[[1]]$taxid_ch1) == length(gse.list[[1]]$geo_accession)) {
    if (unique(gse.list[[1]]$taxid_ch1) == SPECIES) {
      message("PASS: Human samples detected.")
    } else {
      message("FAIL: The samples are not human")
      valid = FALSE
    }
  } else {
    message("FAIL: Data series contains samples from more than one species.")
    valid = FALSE
  }
  
  # Validate beadchip array
  PLATFORMS <-
    c(
      "GPL13534" = "HM450k",
      "GPL16304" = "HM450k",
      "GPL21145" = "EPIC",
      "GPL8490" = "HM27k"
    )
  if (unique(gse.list[[1]]$platform_id) %in% names(PLATFORMS)) {
    message("PASS: Valid platform")
  } else {
    message(
      "FAIL: Only BeadChip methylation arrays 27k, 450k, EPIC and Illumina HiSeq 2500  are currently supported."
    )
    valid = FALSE
  }
  if (table(gse.list[[1]]$platform_id) == length(gse.list[[1]]$geo_accession)) {
    message("PASS: A unique platform ",
            PLATFORMS[gse.list[[1]]@annotation],
            " detected in the data series.")
  } else {
    message("FAIL: More than one platforms have been identified in the data series.")
    valid = FALSE
  }
  return(valid)
}

groupGEO <- function(anno.df, beta.df) {
  options(scipen = 999)
  class.g1 <-
    subset(anno.df, anno.df$status == unique(anno.df$status)[1])
  class.g2 <-
    subset(anno.df, anno.df$status == unique(anno.df$status)[2])
  beta.g1 <- as.data.frame(beta.df[, class.g1[, 1]])
  beta.g2 <- as.data.frame(beta.df[, class.g2[, 1]])
  d.beta <-
    as.data.frame(rowMeans(beta.g1, na.rm = TRUE) - rowMeans(beta.g2, na.rm =
                                                               TRUE))
  colnames(d.beta) <-
    paste0(unique(class.g1$status), "-", unique(class.g2$status))
  db <- cbind(rownames(d.beta), d.beta)
  colnames(db)[1] <- "probes"
  grp1 <- as.data.frame(rowMeans(beta.g1, na.rm = TRUE))
  grp1$probes <- rownames(grp1)
  rownames(grp1) <- NULL
  colnames(grp1) <- c(unique(anno.df$status)[1], "probes")
  grp2 <- as.data.frame(rowMeans(beta.g2, na.rm = TRUE))
  grp2$probes <- rownames(grp2)
  rownames(grp2) <- NULL
  colnames(grp2) <- c(unique(anno.df$status)[2], "probes")
  result <- list()
  result[[1]] <- db
  result[[2]] <- grp1
  result[[3]] <- grp2
  return(result)
}

probesChr <- function(gse, probes.db) {
  if (gse[[1]]@annotation == "GPL8490") {
    hm27 <- as.data.frame(get27k())
    platform.probes <- data.frame(rownames(hm27), hm27[, 1:3])
  } else {
    platform.probes <- cbind(
      gse[[1]]@featureData@data$ID,
      paste0("chr", gse[[1]]@featureData@data$CHR),
      gse[[1]]@featureData@data$MAPINFO,
      gse[[1]]@featureData@data$MAPINFO + 1
    )
  }
  if (ncol(platform.probes) != 4) {
    stop("Data series contains missing data. Exiting...")
  }
  colnames(platform.probes) <- c("probes", "chr", "start", "end")
  bed.data <- merge(probes.db, platform.probes, by = "probes")
  result <- bed.data
  return(result)
}

#' Reformat methylation sequencing data fetched from GEO
#'
#' Transforms sequencing data into bed-formatted files. Valid for per sample usage.
#'
#' @param file.path Local folder of the bed.gz file
#' @param cov Minimum read coverage (default:10)
#' @param chroms A vector of chromosome vector to be included in the analysis (default:ALL)
#' @return A valid bed-formatted data frame
#' @name loadSeqGEO
#' @export
loadSeqGEO <- function(file.path,
                       cov = 10,
                       chroms = NULL) {
  if (!(dplyr::between(cov, 1, 100)))
    stop("Check offset value (valid range [1,100]).")
  file.desc <- gzfile(file.path, open = "r")
  suppressWarnings(gsm <-
                     read.csv(
                       file = file.desc,
                       sep = "\t",
                       header = FALSE
                     ))
  colnames(gsm) = c("chr", "start", "end", "depth", "score", "strand")
  if (is.null(chroms)) {
    chroms = paste0("chr", c(1:22, "X", "Y"))
  }
  f.gsm <- gsm[gsm$chr %in% chroms, ]
  f.gsm$depth <- gsub('[0-9]+/', '', f.gsm$depth)
  f.gsm$depth <- gsub("\\s*\\([^\\)]+\\)", "", f.gsm$depth)
  f.gsm$depth <- as.integer(gsub("\\'", "", f.gsm$depth))
  f.gsm <- f.gsm[f.gsm$depth >= cov, ]
  f.gsm$score <- f.gsm$score / 1000
  f.gsm <- reorderBed(f.gsm, 1, 2, 3, 5, 6)
  close(file.desc)
  return(f.gsm)
}


mval2beta <- function(mval) {
  tr <- 2 ^ (mval)
  beta <- tr / (tr + 1)
  return(beta)
}
